### PR TITLE
New Contact Form After Email Sent Hook

### LINKED ISSUE #8041: Grunion: Add grunion_after_message_sent hook
I'd like to suggest the addition of a `grunion_after_message_sent` hook, in the same manner as the current `grunion_pre_message_sent` runs before a contact for message is sent, there should exist one for after the message is sent.

Use cases would be triggering actions after a successfully sent message (from the Jetpack side of things, as we can't control mailserver behavior of course) which relies on the data originally sent. For example a submission verification to the form submitter. There are plugins that work well if you have a single form on your entire site, but they approach this in awkward workarounds. Having this action would greatly improve the flexibility of contact for mcustomizations.

The format might be something like this:
`do_action( 'grunion_after_message_sent', $to, $message, $headers, $all_values, $extra_values )`

- `$to` is the email address configured in Jetpack to receive the contact form.
- `$message` is the actual message that was sent, with formatting and all.
- `$headers` are the headers supplied for the email.
- `$all_values` and `$extra_values` are the same values as used in the pre-action.

### PR DESCRIPTION
**Pull request by timnolte**
Fixes #8041

#### Changes proposed in this Pull Request:
* Adds new grunion_after_message_sent hook that can be fired after a contact form email has been sent.

#### Testing instructions:
* Add a call to `add_action( 'grunion_after_message_sent', 'my_custom_action' );` in a custom plugin file.
* Add a method `function my_custom_action( $post_id, $to, $subject, $message, $headers, $all_values, $extra_values );`
* Test that all passed parameters contain the expected values.

#### Proposed changelog entry for your changes:
* Adds new Contact Form hook `grunion_after_message_sent` that will fire after a contact form email has been sent.

### CODE REVIEW

**Comment by jeherve** [comment]
Thank you for your contribution!

@clorith Would you like to give this a try, since you were the one to suggest that filter in the first place? Would something like this work for you?

File: modules/contact-form/grunion-contact-form.php
Comment (georgestephanis) [inline modules/contact-form/grunion-contact-form.php]: That's a lotta parameters.  Part of me wonders if we should `compact()` to, subject, message, and headers into one associative array.

Comment (timnolte) [reply modules/contact-form/grunion-contact-form.php]: I will work on this change and update the PR tonight.

**Comment by Clorith** [comment]
@jeherve I'm afraid I don't recall the exact use case any more, but the concept seems fine in and of it self, as it merely replicates the existing `pre_*` behavior.

File: modules/contact-form/grunion-contact-form.php
Comment (timnolte) [reply modules/contact-form/grunion-contact-form.php]: As I was looking for the best way to reduce the number of parameters, as simple `compact()` style of reduction would result in the possible array of `$to` items and the array of `$headers` all being jammed into a single dimension array. Which would also result in a failed/problematic `extract()`. So, while yes it is a seemingly large set of parameters is the best one to give the most flexibility, and provide the most options for post-processing.

**Comment by timnolte** [comment]
What is holding up this PR from being applied/accepted?

**Comment by brbrr** [comment]
Hi @timnolte

There few things: The PR not labeled properly (I'll take care of that) and also one of the CI jobs were failed. I re-launched it. Also, do you think it would be a good idea to add some unit-tests for this new hook?

**Comment by timnolte** [comment]
> Hi @timnolte
> 
> There few things: The PR not labeled properly (I'll take care of that) and also one of the CI jobs were failed. I re-launched it. Also, do you think it would be a good idea to add some unit-tests for this new hook?

OK, yeah, I can work on some unit tests in the contact form module test class.

**Comment by kraftbj** [comment]
D24886-wpcom for corresponding WP.com diff.

@timnolte I bumped the version on your copy of it if you want to pull that change. I'd like to include this in the April release which goes to code freeze the last Tuesday of March.

I'm going to mark this as waiting for author reply if you can get to the unit tests. If not, please let me know! Thank you for your contribution and let me know if there's anything I can do to help!

**Comment by timnolte** [comment]
@kraftbj I'll try to get to those unit tests here in the next week. I've got some fixes for a plugin I'm a co-maintainer on that I'm in the middle of so as soon as I'm done with that release I'll tackle the unit tests. Thanks!

File: modules/contact-form/grunion-contact-form.php
Comment (kraftbj) [reply modules/contact-form/grunion-contact-form.php]: For a bit of prior art, https://developer.wordpress.org/reference/hooks/wp_mail/ compacts those items for the filter, but WP doesn't have a post-send action on `wp_mail` for anything to mimic. There is another hook with a number of parameters too ( https://core.trac.wordpress.org/browser/tags/5.1/src/wp-includes/pluggable.php#L915 ).

I'm fine letting this go in uncompacted.

**Comment by timnolte** [comment]
Working on freeing up some space on my dev machine to get those unit tests done. Targeting the end of the week.

**Review by kraftbj — APPROVED** [review]

**Comment by kraftbj** [comment]
r190883-wpcom

**Pull request merged**
